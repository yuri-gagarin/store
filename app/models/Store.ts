import mongoose, { Schema, Document } from "mongoose";
import { IStorePicture } from "./StorePicture";
type IStorePictureRef = mongoose.Types.ObjectId | IStorePicture;

export interface IStore extends Document {
  title: string;
  description: string;
  images: IStorePictureRef[];
  createdAt: Date;
  editedAt?: Date;
}

const StoreSchema: Schema = new Schema({
  title: {
    type: String,
    required: true
  },
  description: {
    type: String,
    required: true
  },
  images: [
    {
      type: Schema.Types.ObjectId,
      ref: "StorePicture"
    }
  ],
  createdAt: {
    type: Date,
    default: Date.now,
    required: true
  },
  editedAt: {
    type: Date
  }
});
const Store = mongoose.model<IStore>('Store', StoreSchema);
export default Store;