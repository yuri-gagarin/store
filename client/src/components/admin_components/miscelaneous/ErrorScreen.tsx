import React from 'react'
import { Button, Dimmer, Header, Icon, Segment } from 'semantic-ui-react'
import axios from "axios";

type RetryLastReqFunc = <T> () => T
interface Props {
  errorMessage?: string;
  lastRequest?: () => Promise<void>;
}
const ErrorScreen: React.FC<Props> = ({ errorMessage, lastRequest }): JSX.Element => {

  const handleRetry = (e: React.MouseEvent<HTMLButtonElement>): void => {
    if (lastRequest) {
      lastRequest();
    }
  };
  
  return (
    <div style={{ height: "100vh" }}> 
      <Segment style={{ height: "100vh" }}>
        <Dimmer active>
          <Header as='h2' icon inverted>
            <Icon name='exclamation circle' />
              An Error Occured!
            <Header.Subheader>Dimmer sub-header</Header.Subheader>
            <Button 
              id="errorScreenRetryButton"
              inverted 
              color="green" 
              content="Retry" 
              onClick={handleRetry} 
            />
          </Header>
        </Dimmer>
      </Segment>
    </div>
  )
};

export default ErrorScreen