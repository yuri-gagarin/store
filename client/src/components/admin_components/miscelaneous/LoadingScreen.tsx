import React from 'react'
import { Dimmer, Loader, Segment } from 'semantic-ui-react'

const LoaderExampleText: React.FC<{}> = (props): JSX.Element => {
  return (
    <div style={{ height: "100vh" }} id="generalLoadingScreen"> 
      <Segment style={{ height: "100vh" }}>
        <Dimmer active inverted>
          <Loader>Please Wait</Loader>
        </Dimmer>
      </Segment>
    </div>
  )
};

export default LoaderExampleText