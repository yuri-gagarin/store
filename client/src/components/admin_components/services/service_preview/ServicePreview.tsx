import React from "react";
import { Button, Item } from "semantic-ui-react";
// css //
import "./css/servicePreview.css";
// helpers //
import { ConvertDate, setDefaultImage, trimString } from "../../../helpers/displayHelpers";

interface Props {
  service: IServiceData;
}

const ServicePreview: React.FC<Props> = ({ service }): JSX.Element => {
  const { name, description, price, images, createdAt } = service;

  return (
    <Item className="servicePreviewItem">
      <Item.Image size="small" src={setDefaultImage(images)}/>
      <Item.Content>
        <Item.Meta>
          Displayed name: <strong>{name}</strong>
        </Item.Meta>
        <Item.Meta>
          <span className='servicePrice'>Displayed Price: <strong>{price}</strong></span>
          <span className='serviceCreated'>Created At: <strong>{ConvertDate.international(createdAt)}</strong></span>
        </Item.Meta>
        <Item.Description>{trimString(description, 200)}</Item.Description>
        <Button basic color="green" content="View"></Button>
      </Item.Content>
    </Item>
  );
};

export default ServicePreview;