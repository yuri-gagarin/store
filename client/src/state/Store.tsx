import React, { createContext, useReducer } from "react";
import { indexReducer, rootState } from "./reducers/indexReducer";
import { initialStoreState, StoreAction, IStoreState } from "./reducers/storeReducer";
// global app state //
export type IGlobalAppState = {
  storeState: IStoreState;
}
// app actions //
export type AppAction = StoreAction;
// global app context //
export interface IGlobalAppContext {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}
export type CombinedReducer = (state: IGlobalAppState, action: AppAction) => IGlobalAppState;
// context initialization //
const initialContext: IGlobalAppContext = {
  state: {
    storeState: { ...initialStoreState }
  },
  dispatch: (value: AppAction): void => {}
} 
const Store = createContext<IGlobalAppContext>(initialContext);

export const StateProvider: React.FC<{}> = ({ children }): JSX.Element => {
  const [ globalState, dispatch ] = useReducer<CombinedReducer>(indexReducer, rootState as IGlobalAppState);
  return  (
    <Store.Provider value={{ state: globalState, dispatch: dispatch }}>
      {children}
    </Store.Provider>
  );
};

