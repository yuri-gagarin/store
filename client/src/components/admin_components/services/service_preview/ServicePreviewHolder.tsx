import React, { useEffect } from "react";
import { Grid, Item } from "semantic-ui-react";
// additional components //
import ServicePreview from "./ServicePreview";
import ServicesDetails from "./ServicesDetails"
// types and interfaces //
import { AppAction, IGlobalAppState } from "../../../../state/Store";
// api actions //
import { getAllServices } from "../actions/APIServiceActions";
import PopularServiceHolder from "./popular_services/PopularServiceHolder";

interface Props {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}

const ServicePreviewHolder: React.FC<Props> = ({ state, dispatch }): JSX.Element => {
  const { loadedServices } = state.serviceState;
  console.log(loadedServices);
  useEffect(() => {
    getAllServices(dispatch);
  }, []);

  return (
    <Grid stackable padded columns={2}>
      <Grid.Row>
      <Grid.Column computer={10} tablet={8} mobile={16}>
        <Item.Group>
          {
            loadedServices.map((service) => {
              return (
                <ServicePreview 
                  key={service._id}
                  service={service}
                />
              );
            })
          }
        </Item.Group>
      </Grid.Column>
      <Grid.Column computer={6} tablet={8} mobile={16}>
        <ServicesDetails totalServices={loadedServices.length} />
        <PopularServiceHolder popularServices={loadedServices}/>
      </Grid.Column>

      </Grid.Row>
      
    </Grid>
  )
};

export default ServicePreviewHolder;