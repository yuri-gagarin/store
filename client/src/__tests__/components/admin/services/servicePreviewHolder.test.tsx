import React from "react";
import { Grid } from "semantic-ui-react"
import { MemoryRouter as Router } from "react-router-dom";
// test dependencies //
import moxios from "moxios";
import { mount, ReactWrapper } from "enzyme"; 
import { act } from "react-dom/test-utils";
// components //
import ServicePreviewHolder from "../../../../components/admin_components/services/service_preview/ServicePreviewHolder";
import ServicePreview from "../../../../components/admin_components/services/service_preview/ServicePreview";
import LoadingScreen from "../../../../components/admin_components/miscelaneous/LoadingScreen";
import ErrorScreen from "../../../../components/admin_components/miscelaneous/ErrorScreen";
// state //
import { TestStateProvider } from "../../../../state/Store";
// helpers //
import { generateCleanState } from "../../../../test_helpers/miscHelpers";

describe("ServicePreviewHolder Component render tests", () => {
  let services: IServiceData[];

  beforeAll(() => {
    services = [
      {
        _id: "1",
        name: "name",
        price: "100",
        description: "description",
        images: [],
        createdAt: "now"
      },
      {
        _id: "2",
        name: "name",
        price: "200",
        description: "description",
        images: [],
        createdAt: "now"
      }
    ]
  });
  
  // TEST ServicePreviewHolder in its loading state //
  describe("ServicePreviewHolder in 'loading' state", () => {
    let wrapper: ReactWrapper;

    beforeAll( async () => {
      moxios.install();
      wrapper = mount(
        <Router keyLength={0} initialEntries={["/admin/home/my_services/manage"]}>
          <TestStateProvider>
              <ServicePreviewHolder />
          </TestStateProvider>
        </Router>
      );

      await act( async () => {
        await moxios.stubRequest("/api/services", {
          status: 200,
          response: {
            responseMsg: "All ok",
            services: []
          }
        });
      });
    });
    afterAll(() => {
      moxios.uninstall();
    });
    it("Should correctly render", () => {
      expect(wrapper.find(ServicePreviewHolder)).toMatchSnapshot();
    });
    it("Should have a LoadingScreen while 'loading == true'", () => {
      const loadingScreen = wrapper.find(LoadingScreen);
      expect(loadingScreen.length).toEqual(1);
    });
    it("Should NOT display the '#adminServicePreviewHolder' Grid", () => {
      const adminServicePreview = wrapper.find(Grid);
      expect(adminServicePreview.length).toEqual(0);
    }); 
    it("Should NOT display the 'ErrorScreen' component", () => {
      const adminServicePreview = wrapper.find(ErrorScreen);
      expect(adminServicePreview.length).toEqual(0);
    }); 
    it("Should not display any ServicePreview Components", () => {
      const servicePreviewComponents = wrapper.find(ServicePreview);
      expect(servicePreviewComponents.length).toEqual(0);
    });

  });
  
  // END TEST ServicePreviewHolder in its loading state //
  // TEST ServicePreviewHolder in its loaded state //
  describe("ServicePreview in 'loaded' state", () => {
    let wrapper: ReactWrapper;

    beforeAll( async () => {
      moxios.install();
      const mockState = generateCleanState();

      wrapper = mount(
        <Router keyLength={0} initialEntries={["/admin/home/my_services/manage"]}>
          <TestStateProvider mockState={mockState}>
            <ServicePreviewHolder />
          </TestStateProvider>
         </Router>
      );

      await act( async () => {
        await moxios.stubRequest("/api/services", {
          status: 200,
          response: {
            responseMsg: "All ok",
            services: services
          }
        });
      });
      moxios.uninstall();
    });
    
    it("Should correctly render initial 'LoadingScreen' component", () => {
      const loadingScreen = wrapper.find(LoadingScreen);
      const errorScreen = wrapper.find(ErrorScreen);
      const serviceGrid = wrapper.find(Grid);
      // assert correct rendering //
      expect(loadingScreen.length).toEqual(1)
      expect(errorScreen.length).toEqual(0);
      expect(serviceGrid.length).toEqual(0);
    });
    it("Should correctlt render the '#adminServicePreviewHolder' Grid", () => {
      wrapper.update();
      const loadingScreen = wrapper.find(LoadingScreen);
      const errorScreen = wrapper.find(ErrorScreen);
      const serviceGrid = wrapper.find(Grid);
      // assert correct rendering ///
      expect(loadingScreen.length).toEqual(0);
      expect(errorScreen.length).toEqual(0);
      expect(serviceGrid.length).toEqual(1);
      expect(wrapper.find(ServicePreviewHolder)).toMatchSnapshot();
    });
    it(`Should display a correct number of ServicePreview Components`, () => {
      const servicePreviewComponents = wrapper.find(ServicePreview);
      expect(servicePreviewComponents.length).toEqual(2);
    });
  });
  
  // END TEST ServicePreviewHolder in its loaded state //
  // TEST ServicePreview Component in Error state //
  
  describe("ServicePreview Component in Error state", () => {
    let wrapper: ReactWrapper;
    const error = new Error("Error occured");

    beforeAll(async () => {
      const mockState = generateCleanState();

      await act( async () => {
        moxios.install();
        wrapper = mount(
          <Router keyLength={0} initialEntries={["/admin/home/my_services/view_all"]}>
            <TestStateProvider mockState={mockState}>
              <ServicePreviewHolder />
            </TestStateProvider>
          </Router>
        );
        moxios.stubRequest("/api/services", {
          status: 500,
          response: {
            responseMsg: "An error occured",
            error: error
          }
        })
      });
      moxios.uninstall();
    })
    afterAll(() => {
      moxios.uninstall();
    });

    it("Should initially render the 'LoadingScreen' component", () => {
      const loadingScreen = wrapper.find(ServicePreviewHolder).find(LoadingScreen);
      const errorScreen = wrapper.find(ServicePreviewHolder).find(ErrorScreen);
      const servicesGrid = wrapper.find(ServicePreviewHolder).find(Grid);
      // assert correct rendering //
      expect(loadingScreen.length).toEqual(1);
      expect(errorScreen.length).toEqual(0);
      expect(servicesGrid.length).toEqual(0);
    });
    
    it("Should initially render the 'ErrorScreen' component after rerender", () => {
      wrapper.update();
      const loadingScreen = wrapper.find(ServicePreviewHolder).find(LoadingScreen);
      const errorScreen = wrapper.find(ServicePreviewHolder).find(ErrorScreen);
      const servicesGrid = wrapper.find(ServicePreviewHolder).find(Grid);
      // assert correct rendering //
      expect(loadingScreen.length).toEqual(0);
      expect(errorScreen.length).toEqual(1);
      expect(servicesGrid.length).toEqual(0);
    });
    it("Should correctly respond to '#errorScreenRetryBtn' and rerender", async () => {
      await act( async () => {
        moxios.install();
        moxios.stubRequest("/api/services", {
          status: 200,
          response: {
            responseMsg: "All ok",
            services: services
          }
        })
        const retryButton = wrapper.find("#errorScreenRetryButton");
        retryButton.at(0).simulate("click");
      });
      
    });
  });
  
});