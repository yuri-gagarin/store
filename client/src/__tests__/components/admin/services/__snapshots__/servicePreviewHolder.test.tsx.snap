// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServicePreviewHolder Component render tests ServicePreview in 'loaded' state Should correctlt render the '#adminServicePreviewHolder' Grid 1`] = `
<withRouter(ServicePreviewHolder)>
  <ServicePreviewHolder
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/admin/home/my_services/manage",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/admin/home/my_services/manage",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    location={
      Object {
        "hash": "",
        "pathname": "/admin/home/my_services/manage",
        "search": "",
        "state": undefined,
      }
    }
    match={
      Object {
        "isExact": false,
        "params": Object {},
        "path": "/",
        "url": "/",
      }
    }
  >
    <Grid
      columns={2}
      id="adminServicePreviewHolder"
      padded={true}
      stackable={true}
    >
      <div
        className="ui stackable padded two column grid"
        id="adminServicePreviewHolder"
      >
        <GridRow>
          <div
            className="row"
          >
            <GridColumn
              computer={10}
              mobile={16}
              tablet={8}
            >
              <div
                className="ten wide computer sixteen wide mobile eight wide tablet column"
              >
                <ItemGroup>
                  <div
                    className="ui items"
                  >
                    <ServicePreview
                      key="1"
                      service={
                        Object {
                          "_id": "1",
                          "createdAt": "now",
                          "description": "description",
                          "images": Array [],
                          "name": "name",
                          "price": "100",
                        }
                      }
                    >
                      <Item
                        className="servicePreviewItem"
                      >
                        <div
                          className="item servicePreviewItem"
                        >
                          <ItemImage
                            size="small"
                            src="https://react.semantic-ui.com/images/wireframe/image.png"
                          >
                            <Image
                              as="img"
                              size="small"
                              src="https://react.semantic-ui.com/images/wireframe/image.png"
                              ui={true}
                              wrapped={true}
                            >
                              <div
                                className="ui small image"
                              >
                                <img
                                  src="https://react.semantic-ui.com/images/wireframe/image.png"
                                />
                              </div>
                            </Image>
                          </ItemImage>
                          <ItemContent>
                            <div
                              className="content"
                            >
                              <ItemMeta>
                                <div
                                  className="meta"
                                >
                                  Displayed name: 
                                  <strong>
                                    name
                                  </strong>
                                </div>
                              </ItemMeta>
                              <ItemMeta>
                                <div
                                  className="meta"
                                >
                                  <span
                                    className="servicePrice"
                                  >
                                    Displayed Price: 
                                    <strong>
                                      100
                                    </strong>
                                  </span>
                                  <span
                                    className="serviceCreated"
                                  >
                                    Created At: 
                                    <strong>
                                      NaN-undefined-NaN
                                    </strong>
                                  </span>
                                </div>
                              </ItemMeta>
                              <ItemDescription>
                                <div
                                  className="description"
                                >
                                  description...
                                </div>
                              </ItemDescription>
                              <Button
                                as="button"
                                basic={true}
                                color="green"
                                content="View"
                              >
                                <Ref
                                  innerRef={
                                    Object {
                                      "current": <button
                                        class="ui green basic button"
                                      >
                                        View
                                      </button>,
                                    }
                                  }
                                >
                                  <RefFindNode
                                    innerRef={
                                      Object {
                                        "current": <button
                                          class="ui green basic button"
                                        >
                                          View
                                        </button>,
                                      }
                                    }
                                  >
                                    <button
                                      className="ui green basic button"
                                      onClick={[Function]}
                                    >
                                      View
                                    </button>
                                  </RefFindNode>
                                </Ref>
                              </Button>
                            </div>
                          </ItemContent>
                        </div>
                      </Item>
                    </ServicePreview>
                    <ServicePreview
                      key="2"
                      service={
                        Object {
                          "_id": "2",
                          "createdAt": "now",
                          "description": "description",
                          "images": Array [],
                          "name": "name",
                          "price": "200",
                        }
                      }
                    >
                      <Item
                        className="servicePreviewItem"
                      >
                        <div
                          className="item servicePreviewItem"
                        >
                          <ItemImage
                            size="small"
                            src="https://react.semantic-ui.com/images/wireframe/image.png"
                          >
                            <Image
                              as="img"
                              size="small"
                              src="https://react.semantic-ui.com/images/wireframe/image.png"
                              ui={true}
                              wrapped={true}
                            >
                              <div
                                className="ui small image"
                              >
                                <img
                                  src="https://react.semantic-ui.com/images/wireframe/image.png"
                                />
                              </div>
                            </Image>
                          </ItemImage>
                          <ItemContent>
                            <div
                              className="content"
                            >
                              <ItemMeta>
                                <div
                                  className="meta"
                                >
                                  Displayed name: 
                                  <strong>
                                    name
                                  </strong>
                                </div>
                              </ItemMeta>
                              <ItemMeta>
                                <div
                                  className="meta"
                                >
                                  <span
                                    className="servicePrice"
                                  >
                                    Displayed Price: 
                                    <strong>
                                      200
                                    </strong>
                                  </span>
                                  <span
                                    className="serviceCreated"
                                  >
                                    Created At: 
                                    <strong>
                                      NaN-undefined-NaN
                                    </strong>
                                  </span>
                                </div>
                              </ItemMeta>
                              <ItemDescription>
                                <div
                                  className="description"
                                >
                                  description...
                                </div>
                              </ItemDescription>
                              <Button
                                as="button"
                                basic={true}
                                color="green"
                                content="View"
                              >
                                <Ref
                                  innerRef={
                                    Object {
                                      "current": <button
                                        class="ui green basic button"
                                      >
                                        View
                                      </button>,
                                    }
                                  }
                                >
                                  <RefFindNode
                                    innerRef={
                                      Object {
                                        "current": <button
                                          class="ui green basic button"
                                        >
                                          View
                                        </button>,
                                      }
                                    }
                                  >
                                    <button
                                      className="ui green basic button"
                                      onClick={[Function]}
                                    >
                                      View
                                    </button>
                                  </RefFindNode>
                                </Ref>
                              </Button>
                            </div>
                          </ItemContent>
                        </div>
                      </Item>
                    </ServicePreview>
                  </div>
                </ItemGroup>
              </div>
            </GridColumn>
            <GridColumn
              computer={6}
              mobile={16}
              tablet={8}
            >
              <div
                className="six wide computer sixteen wide mobile eight wide tablet column"
              >
                <ServicesDetails
                  totalServices={2}
                >
                  <div
                    className="adminServiceDetailsHolder"
                  >
                    <div
                      className="adminServiceDetailsTitle"
                    >
                      Service Details
                    </div>
                    <div
                      className="adminServiceDetails"
                    >
                      <div
                        className="adminServiceCounter"
                      >
                        <div>
                          Total Services:
                        </div>
                        <div
                          className="adminServicesCount"
                        >
                          2
                        </div>
                      </div>
                    </div>
                  </div>
                </ServicesDetails>
                <PopularServiceHolder
                  popularServices={
                    Array [
                      Object {
                        "_id": "1",
                        "createdAt": "now",
                        "description": "description",
                        "images": Array [],
                        "name": "name",
                        "price": "100",
                      },
                      Object {
                        "_id": "2",
                        "createdAt": "now",
                        "description": "description",
                        "images": Array [],
                        "name": "name",
                        "price": "200",
                      },
                    ]
                  }
                >
                  <div
                    className="popularServiceHolder"
                  >
                    <div
                      className="popServiceHolTitle"
                    >
                      Popular Services
                    </div>
                    <ItemGroup
                      divided={true}
                    >
                      <div
                        className="ui divided items"
                      >
                        <PopularService
                          key="1"
                          popularService={
                            Object {
                              "_id": "1",
                              "createdAt": "now",
                              "description": "description",
                              "images": Array [],
                              "name": "name",
                              "price": "100",
                            }
                          }
                        >
                          <Item>
                            <div
                              className="item"
                            >
                              <ItemImage
                                size="tiny"
                                src="https://react.semantic-ui.com/images/wireframe/image.png"
                              >
                                <Image
                                  as="img"
                                  size="tiny"
                                  src="https://react.semantic-ui.com/images/wireframe/image.png"
                                  ui={true}
                                  wrapped={true}
                                >
                                  <div
                                    className="ui tiny image"
                                  >
                                    <img
                                      src="https://react.semantic-ui.com/images/wireframe/image.png"
                                    />
                                  </div>
                                </Image>
                              </ItemImage>
                              <ItemContent>
                                <div
                                  className="content"
                                >
                                  <ItemHeader>
                                    <div
                                      className="header"
                                    >
                                      name
                                    </div>
                                  </ItemHeader>
                                  <ItemMeta>
                                    <div
                                      className="meta"
                                    >
                                      <span>
                                        Priced At: 
                                        <strong>
                                          100
                                        </strong>
                                      </span>
                                      <span>
                                        -
                                      </span>
                                      <span>
                                        Posted: 
                                        <strong>
                                          NaN-undefined-NaN
                                        </strong>
                                      </span>
                                    </div>
                                  </ItemMeta>
                                  <Button
                                    as="button"
                                    basic={true}
                                    color="blue"
                                    content="More..."
                                  >
                                    <Ref
                                      innerRef={
                                        Object {
                                          "current": <button
                                            class="ui blue basic button"
                                          >
                                            More...
                                          </button>,
                                        }
                                      }
                                    >
                                      <RefFindNode
                                        innerRef={
                                          Object {
                                            "current": <button
                                              class="ui blue basic button"
                                            >
                                              More...
                                            </button>,
                                          }
                                        }
                                      >
                                        <button
                                          className="ui blue basic button"
                                          onClick={[Function]}
                                        >
                                          More...
                                        </button>
                                      </RefFindNode>
                                    </Ref>
                                  </Button>
                                </div>
                              </ItemContent>
                            </div>
                          </Item>
                        </PopularService>
                        <PopularService
                          key="2"
                          popularService={
                            Object {
                              "_id": "2",
                              "createdAt": "now",
                              "description": "description",
                              "images": Array [],
                              "name": "name",
                              "price": "200",
                            }
                          }
                        >
                          <Item>
                            <div
                              className="item"
                            >
                              <ItemImage
                                size="tiny"
                                src="https://react.semantic-ui.com/images/wireframe/image.png"
                              >
                                <Image
                                  as="img"
                                  size="tiny"
                                  src="https://react.semantic-ui.com/images/wireframe/image.png"
                                  ui={true}
                                  wrapped={true}
                                >
                                  <div
                                    className="ui tiny image"
                                  >
                                    <img
                                      src="https://react.semantic-ui.com/images/wireframe/image.png"
                                    />
                                  </div>
                                </Image>
                              </ItemImage>
                              <ItemContent>
                                <div
                                  className="content"
                                >
                                  <ItemHeader>
                                    <div
                                      className="header"
                                    >
                                      name
                                    </div>
                                  </ItemHeader>
                                  <ItemMeta>
                                    <div
                                      className="meta"
                                    >
                                      <span>
                                        Priced At: 
                                        <strong>
                                          200
                                        </strong>
                                      </span>
                                      <span>
                                        -
                                      </span>
                                      <span>
                                        Posted: 
                                        <strong>
                                          NaN-undefined-NaN
                                        </strong>
                                      </span>
                                    </div>
                                  </ItemMeta>
                                  <Button
                                    as="button"
                                    basic={true}
                                    color="blue"
                                    content="More..."
                                  >
                                    <Ref
                                      innerRef={
                                        Object {
                                          "current": <button
                                            class="ui blue basic button"
                                          >
                                            More...
                                          </button>,
                                        }
                                      }
                                    >
                                      <RefFindNode
                                        innerRef={
                                          Object {
                                            "current": <button
                                              class="ui blue basic button"
                                            >
                                              More...
                                            </button>,
                                          }
                                        }
                                      >
                                        <button
                                          className="ui blue basic button"
                                          onClick={[Function]}
                                        >
                                          More...
                                        </button>
                                      </RefFindNode>
                                    </Ref>
                                  </Button>
                                </div>
                              </ItemContent>
                            </div>
                          </Item>
                        </PopularService>
                      </div>
                    </ItemGroup>
                  </div>
                </PopularServiceHolder>
              </div>
            </GridColumn>
          </div>
        </GridRow>
      </div>
    </Grid>
  </ServicePreviewHolder>
</withRouter(ServicePreviewHolder)>
`;

exports[`ServicePreviewHolder Component render tests ServicePreviewHolder in 'loading' state Should correctly render 1`] = `
<withRouter(ServicePreviewHolder)>
  <ServicePreviewHolder
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/admin/home/my_services/manage",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/admin/home/my_services/manage",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    location={
      Object {
        "hash": "",
        "pathname": "/admin/home/my_services/manage",
        "search": "",
        "state": undefined,
      }
    }
    match={
      Object {
        "isExact": false,
        "params": Object {},
        "path": "/",
        "url": "/",
      }
    }
  >
    <LoaderExampleText>
      <div
        id="generalLoadingScreen"
        style={
          Object {
            "height": "100vh",
          }
        }
      >
        <Segment
          style={
            Object {
              "height": "100vh",
            }
          }
        >
          <div
            className="ui segment"
            style={
              Object {
                "height": "100vh",
              }
            }
          >
            <Dimmer
              active={true}
              inverted={true}
            >
              <DimmerInner
                active={true}
                inverted={true}
              >
                <Ref
                  innerRef={
                    Object {
                      "current": null,
                    }
                  }
                >
                  <RefFindNode
                    innerRef={
                      Object {
                        "current": null,
                      }
                    }
                  >
                    <div
                      className="ui active transition visible inverted dimmer"
                      onClick={[Function]}
                    >
                      <div
                        className="content"
                      >
                        <Loader>
                          <div
                            className="ui text loader"
                          >
                            Please Wait
                          </div>
                        </Loader>
                      </div>
                    </div>
                  </RefFindNode>
                </Ref>
              </DimmerInner>
            </Dimmer>
          </div>
        </Segment>
      </div>
    </LoaderExampleText>
  </ServicePreviewHolder>
</withRouter(ServicePreviewHolder)>
`;
